### PR TITLE
WIP: proposal to implement a subscription factory with relative offset.

### PR DESCRIPTION
**Pull request by jewertow**
## Purpose

Introduction of a subscription factory with relative offset definitions.

## References

See #538

## Changes

- Add `RelativeOffsetSubscriptions.assignmentForLastN`

## Background Context

There has been a need to query certain Kafka topics with relative offsets, for example last offset, or last offset minus 100, or last offset minus 10 minutes.

### CODE REVIEW

**Comment by lightbend-cla-validator** [comment]
Hi @jewertow,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

**Comment by jewertow** [comment]
@ennru I will be grateful if you tell me if this is a right way.

**Comment by ennru** [comment]
Hi @jewertow 	
Thank you for contributing to Alpakka Kafka!
The implementation needs to get more involved as this so that it can be sent created without a reference to the actor. It would introduce new sub-classes to `SubscriptionRequest` that are sent to the actor which internally does the math.

**Comment by jewertow** [comment]
I will continue this PR when I finish #900.

**Comment by jewertow** [comment]
@ennru 
For now I have broaden my knowledge about the project so I prepared new proposal which is only a rough idea of target implementation. Please check it out and give me a feedback.

**Comment by ennru** [comment]
Thank you for getting back to this.

I think it would be interesting to explore how the new [PartitionAssignmentHandler](https://doc.akka.io/docs/alpakka-kafka/snapshot/consumer-rebalance.html#partition-assignment-handler) can be used to implement any kind of non-standard seeking. The idea would be to provide a handler which would react on assignments and apply custom seeking via the provided `RestrictedConsumer`.

File: core/src/main/scala/akka/kafka/scaladsl/RelativeOffsetAssignmentHandler.scala [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+package akka.kafka.scaladsl
+
+import akka.kafka.RestrictedConsumer
+import org.apache.kafka.common.TopicPartition
+
+import scala.jdk.CollectionConverters._
+
+object RelativeOffsetAssignmentHandler {
+  def apply(backOffset: Long): RelativeOffsetAssignmentHandler =
```
Comment (jewertow) [inline core/src/main/scala/akka/kafka/scaladsl/RelativeOffsetAssignmentHandler.scala]: It seeks all assigned partitions to offset = endOffset - backOffset.

File: core/src/main/scala/akka/kafka/scaladsl/RelativeOffsetAssignmentHandler.scala [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    new RelativeOffsetAssignmentHandler(None, Some(backOffset))
+
+  def apply(topicPartition: TopicPartition, backOffset: Long): RelativeOffsetAssignmentHandler =
```
Comment (jewertow) [inline core/src/main/scala/akka/kafka/scaladsl/RelativeOffsetAssignmentHandler.scala]: It seeks given partition to offset = end offset - backOffset and all others are assigned to beginning offset.

File: core/src/main/scala/akka/kafka/scaladsl/RelativeOffsetAssignmentHandler.scala [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    new RelativeOffsetAssignmentHandler(Some(Map(topicPartition -> backOffset)))
+
+  def apply(topicPartitionBackOffset: Map[TopicPartition, Long]): RelativeOffsetAssignmentHandler =
```
Comment (jewertow) [inline core/src/main/scala/akka/kafka/scaladsl/RelativeOffsetAssignmentHandler.scala]: It allows to seek multiple partitions to offset = end offset - specified offset. All others assigned partitions are seeking to beginning offset.

File: core/src/main/scala/akka/kafka/scaladsl/RelativeOffsetAssignmentHandler.scala [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    new RelativeOffsetAssignmentHandler(Some(topicPartitionBackOffset))
+
+  def apply(tps: Set[TopicPartition], backOffset: Long): RelativeOffsetAssignmentHandler = {
```
Comment (jewertow) [inline core/src/main/scala/akka/kafka/scaladsl/RelativeOffsetAssignmentHandler.scala]: This method is intended to seek back all the given partitions with the same offset. All others assigned partitions are seeking to beginning offset.

**Comment by jewertow** [comment]
@ennru 
I couldn't work on it recently, but I get back to the issue and I proposed new implementation based on PartitionAssignmentHandler. Please check it out.

**Comment by jewertow** [comment]
@seglo what do you think about it?

**Comment by ennru** [comment]
I would expect this relative offset seeking is interesting only when a partition is assigned the first time after an Alpakka Kafka consumer is started. 
With what you suggest right now it would be applied even when there is a regular rebalance which "brings back" a partition to the consumer.

**Comment by jewertow** [comment]
I forgot about rebalances. I will investigate it again.

**Comment by seglo** [comment]
@jewertow Just curious why you closed the PR.  Are you still interested in pursuing this?

**Comment by jewertow** [comment]
Yes, I want to implement it in the near future, but first I want to finish few other tasks in other projects that I opened. Open but inactive PR is annoying and disrupts the status of work in the project.

**Comment by seglo** [comment]
Alright.  Hope to see you back soon :)

**Pull request closed without merging**